# CINECA_synthetic_cohort_EUROPE_UK README

We have included the [CINECA_synthetic_cohort_EUROPE_UK1](https://www.cineca-project.eu/cineca-synthetic-datasets) dataset to test B2RI functionalities.

The dataset is part of a [study](https://ega-archive.org/studies/EGAS00001002472) consisting of two datasets.

## Dataset description (EGAD00001006673):
The dataset consists of 2504 samples which have genetic data (low coverage WGS) based on [1000 Genomes](https://www.nature.com/articles/nature15393) data (phase3 and Geuvadis), and 76 synthetic subject attributes and phenotypic data derived from UKBiobank. The purpose of this dataset is to aid development of technical implementations for cohort data discovery, harmonization, access, and federated analysis. In support of FAIRness in data sharing, this dataset is made freely available under the Creative Commons Licence (CC-BY). Please ensure this preamble is included with this dataset and that the CINECA project (funding: EC H2020 grant 825775) is acknowledged.

The dataset `EGAD00001006673` was downloaded from the EGA, please see the full description [here](https://ega-archive.org/datasets/EGAD00001006673). Please contact `helpdesk@ega-archive.org` should you want get access to the whole dataset.

A [raw file](./uk1.tsv) with phenoclinic data was gently donated to us from our CINECA-EU partners. 

 
### Notes about column mapping
* `individualId` was taken from the column "Subject Id" (e.g., HG00096) in the [raw data](./uk1.tsv), which in this particular dataset also matches the `biosampleId` in the `VCF`. An alternative way of labelling `individualId` would be by using the column "eid" (e.g., fake1) (not used).

* Note that some terms/variables related to `analyses` and `runs` were not present, nor they could be easily extracted (and matched to samples) from [10000 Genomes phase 3 release](https://www.internationalgenome.org/data-portal/data-collection/phase-3). Thus, some ids (e.g., `runId`, `analysisId`) were created _ad hoc_ for demonstration purposes. 

### Files included:

* `uk1.tsv` - Raw file with the metadata/phenoclinic data for 2504 fake individuals.
* `Beacon-v2-Models_CINECA_UK1.xlsx` - Excel file with the metadata/phenoclinic data (INPUT).
* `bff/*json` - Collections (JSON arrays) created from the Excel file (OUTPUT).
* `scripts/` - Directory containing miscellaneous (ad hoc) scripts utilized by the author for converting `uk1.tsv` into multiple `csv` files, which were then used to populate `Beacon-v2-Models_CINECA_UK1.xlsx`. The primary purpose of these scripts is to demonstrate the methods used for data transformation. It's important to note that **these scripts are tailored to the author's specific data set and are not directly applicable to other data sets. Users are encouraged to develop their own ETL methods for transforming raw data into the CSV format**.

## External Files (Google Drive Downloads)

#### INPUT

- **VCF file:** `chr22.Test.1000G.phase3.joint.vcf.gz` - Whole Genome Sequencing (WGS) data for chromosome 22.
  - [Link to VCF file](https://drive.google.com/file/d/1lcsl87cp3zr6iOlm3VR1QaU_58maX3Z4/view?usp=sharing)
  - [Link to VCF index (.tbi)](https://drive.google.com/file/d/1fkFAmc9iODFgxB0-L6UyS2MAl8YRhuG9/view?usp=sharing)

- **Parameter file:** Use the following YAML configuration and save it as `param.yaml`:

```yaml
genome: hs37
bff2html: false
datasetid: CINECA_synthetic_cohort_EUROPE_UK1
projectdir: cineca_uk1
```

- **Run the analysis** (approx. 25 minutes):

```bash
../bin/bff-tools vcf -i chr22.Test.1000G.phase3.joint.vcf.gz -p param.yaml -t 4
```

#### OUTPUT

For reference, you can also download the output file generated by a previous run.

- **Genomic Variations Collection:** `genomicVariationsVcf.json.gz`
  - [Link to BFF File](https://drive.google.com/file/d/1D8CvBAO1EJzF7m7R0XzqL4AcYrVYgNwl/view?usp=sharing)
