
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Beacon v2 CBI Tools Documentation">
      
      
        <meta name="author" content="Manuel Rueda, PhD">
      
      
      
        <link rel="prev" href="../download-and-installation/non-containerized/">
      
      
        <link rel="next" href="../examples/hg38/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.9">
    
    
      
        <title>üöÄ Quick Start - Beacon v2 CBI Tools Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/theme_overrides.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Beacon v2 CBI Tools Documentation" class="md-header__button md-logo" aria-label="Beacon v2 CBI Tools Documentation" data-md-component="logo">
      
  <img src="../img/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Beacon v2 CBI Tools Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              üöÄ Quick Start
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="amber"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/mrueda/beacon2-cbi-tools" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    beacon2-cbi-tools
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Beacon v2 CBI Tools Documentation" class="md-nav__button md-logo" aria-label="Beacon v2 CBI Tools Documentation" data-md-component="logo">
      
  <img src="../img/logo.png" alt="logo">

    </a>
    Beacon v2 CBI Tools Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/mrueda/beacon2-cbi-tools" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    beacon2-cbi-tools
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üè† Introduction
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    ‚¨áÔ∏è  Download & Installation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            ‚¨áÔ∏è  Download & Installation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../download-and-installation/docker-based/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üê≥ Docker-based
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../download-and-installation/non-containerized/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üõ†Ô∏è Non-containerized
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    üöÄ Quick Start
    
  </span>
  

      </a>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    üß™ Examples
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            üß™ Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/hg38/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üß¨ GRCh38 / hg38
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../data-beaconization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üìö Tutorial
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/CNAG-Biomedical-Informatics/beacon2-cbi-tools/tree/main/CINECA_synthetic_cohort_EUROPE_UK1" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üìä CINECA cohort
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/CNAG-Biomedical-Informatics/beacon2-cbi-tools/blob/main/utils/README.md" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üõ†Ô∏è  Utilities
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="">
            
  
  <span class="md-ellipsis">
    üÜò Help
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            üÜò Help
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../help/faq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ‚ùì FAQ
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../help/troubleshooting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üßØ Troubleshooting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="">
            
  
  <span class="md-ellipsis">
    ‚ÑπÔ∏è About
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            ‚ÑπÔ∏è About
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/about/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üìÑ About
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/citation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    üìñ Citation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="quick-start-bff-tools">Quick Start <code>bff-tools</code><a class="headerlink" href="#quick-start-bff-tools" title="Permanent link">&para;</a></h1>
<p>This guide provides a one-page cheat sheet for common commands:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="c1"># Display help for the tool</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>bin/bff-tools<span class="w"> </span>--help
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="c1"># Display man for the tool</span>
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>bin/bff-tools<span class="w"> </span>--man
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="c1"># Convert VCF to BFF</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>bin/bff-tools<span class="w"> </span>vcf<span class="w"> </span>-i<span class="w"> </span>test/test_1000G.vcf.gz<span class="w"> </span>-p<span class="w"> </span>test/param.yaml
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="c1"># Validate metadata and convert to BFF</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>mkdir<span class="w"> </span>bff_out
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>bin/bff-tools<span class="w"> </span>validate<span class="w"> </span>-i<span class="w"> </span>Beacon-v2-Models_CINECA_UK1.xlsx<span class="w"> </span>--out-dir<span class="w"> </span>bff_out
</span></code></pre></div>
<h1 id="usage">Usage<a class="headerlink" href="#usage" title="Permanent link">&para;</a></h1>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Full Usage</label><label for="__tabbed_1_2">Notes on <code>bff-tool validate</code></label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p># NAME</p>
<p><code>bff-tools</code>: A unified command-line toolkit for working with Beacon v2 Models data. It allows users to <strong>annotate</strong> and <strong>convert</strong> VCF files into the <code>genomicVariations</code> entity using the Beacon-Friendly Format (BFF), <strong>validate</strong> metadata files (XLSX or JSON) against Beacon v2 schema definitions and <strong>load</strong> BFF-formatted data into a <strong>MongoDB</strong> instance.</p>
<p>This tool is part of the <code>beacon2-cbi-tools</code> repository and is designed to support Beacon v2 data ingestion pipelines, metadata validation workflows, and federated data sharing initiatives.</p>
<p># SYNOPSIS</p>
<p>bff-tools &lt;mode&gt; [-arguments] [-options]</p>
<div class="language-text highlight"><pre><span></span><code> Mode:
   vcf
      -i | --input &lt;file&gt;            Requires a VCF.gz file
                                     (May also use a parameters file)

   load
                                     (Requires a parameters file specifying BFF files)

    full (vcf + load)
      -i | --input &lt;file&gt;            Requires a VCF.gz file
                                     (May also use a parameters file)

      Options (for vcf / load / full):
      -c | --config &lt;file&gt;           Requires a configuration file
      -p | --param &lt;file&gt;            Requires a parameters file (optional)
      -projectdir-override &lt;path&gt;    Custom project directory path (overrides config)
      -t | --threads &lt;number&gt;        Number of threads (optional, mainly for VCF)

   validate
      -i | --input &lt;file(s)&gt;         One or more XLSX/JSON metadata files
      -s | --schema-dir &lt;directory&gt;  Directory containing JSON schemas
      -o | --out-dir &lt;directory&gt;     Output directory for validated data
      -gv                            Set this option if you want to process &lt;genomicVariations&gt; entity
      -ignore-validation             Writes JSON collection regardless of results from validation against JSON schemas (AYOR!)

      Experimental:
      -gv-vcf                        Set this option to read &lt;genomicVariations.json&gt; from &lt;beacon vcf&gt; (with one document per line)

    Generic Options:
      -h                             Brief help message
      -man                           Full documentation
      -v                             Display version information
      -debug &lt;level&gt;                 Print debugging information (1 to 5)
      -verbose                       Enable verbosity
      -nc | --no-color               Do not print colors to STDOUT
</code></pre></div>
<p># DESCRIPTION</p>
<p>### <code>bff-tools</code></p>
<p><code>bff-tools</code> is a command-line toolkit with four operational modes for working with Beacon v2 data:</p>
<p># HOW TO RUN <code>bff-tools</code></p>
<p>This script supports four <strong>modes</strong>: <code>vcf</code>, <code>load</code>, <code>full</code>, and <code>validate</code>.</p>
<p><strong>* Mode <code>vcf</code></strong></p>
<p>Annotates a gzipped VCF file (optional) and serializes it into the Beacon-Friendly Format (BFF) as <code>genomicVariationsVcf.json.gz</code>.</p>
<p><strong>* Mode <code>load</code></strong></p>
<p>Loads BFF-formatted JSON files - including metadata and genomic variations - into a MongoDB instance.</p>
<p><strong>* Mode <code>full</code></strong></p>
<p>Combines <code>vcf</code> and <code>load</code>: it processes a VCF file and ingests the resulting data into MongoDB.</p>
<p><strong>* Mode <code>validate</code></strong></p>
<p>Validates metadata files (XLSX or JSON) against the Beacon v2 schema definitions and serializes them into BFF JSON collections.<br />
 Note: This mode uses a separate internal script and does not require a parameters or configuration file.</p>
<p>To perform these tasks, you may need:</p>
<ul>
<li>
<p>A gzipped VCF file (for modes: <code>vcf</code> and <code>full</code>)</p>
<p>Note: It does not need to be bgzipped.</p>
</li>
<li>
<p>A parameters file (optional)</p>
<p>YAML file with job-specific values and metadata file references. Recommended for structured processing.</p>
</li>
<li>
<p>BFF JSON files (required for modes: <code>load</code> and <code>full</code>)</p>
<p>See <a href="#what-is-the-beacon-friendly-format-bff">Beacon-Friendly Format (BFF)</a> for a detailed explanation.</p>
</li>
<li>
<p>Metadata files (XLSX or JSON) (for mode: <code>validate</code>)</p>
<p>You can start with the provided Excel template and use <code>--gv</code> or <code>--ignore-validation</code> flags if needed.</p>
</li>
<li>
<p>Threads (only for <code>vcf</code> and <code>full</code> modes)</p>
<p>You can set the number of threads using <code>-t</code>. However, since SnpEff doesn't parallelize efficiently, it's best to use <code>-t 1</code> and distribute the work (e.g., by chromosome) using GNU <code>parallel</code> or the included <a href="https://github.com/CNAG-Biomedical-Informatics/beacon2-cbi-tools/tree/main/utils/bff_queue">queue system</a>).</p>
</li>
</ul>
<p><code>bff-tools</code> will create an independent project directory <code>projectdir</code> and store all needed information needed there. Thus, many concurrent calculations are supported.
 Note that <code>bff-tools</code> will treat your data as <em>read-only</em> (i.e., will not modify your original files).</p>
<p><strong>Annex: Parameters file</strong> (YAML)</p>
<p>Example for <code>vcf</code> mode:</p>
<div class="language-text highlight"><pre><span></span><code> --
 genome: hs37 # default hg19
 annotate: true # default true
 bff2html: true # default false
</code></pre></div>
<p>Example for <code>load</code> mode:</p>
<div class="language-text highlight"><pre><span></span><code> --
 bff:
   metadatadir: .
   analyses: analyses.json
   biosamples: biosamples.json
   cohorts: cohorts.json
   datasets: datasets.json
   individuals: individuals.json
   runs: runs.json
   # Note that genomicVariationsVcf is not affected by &lt;metadatadir&gt;
   genomicVariationsVcf: beacon_XXXX/vcf/genomicVariationsVcf.json.gz
 projectdir: my_project
</code></pre></div>
<p>Example for <code>full</code> mode:</p>
<div class="language-text highlight"><pre><span></span><code> --
 genome: hs37 # default hg19
 annotate: true # default true
 bff:
   metadatadir: .
   analyses: analyses.json
   runs: runs.json
 projectdir: my_project
</code></pre></div>
<p>Please find below a detailed description of all parameters (alphabetical order):</p>
<ul>
<li>
<p><strong>annotate</strong></p>
<p>When the <strong>annotate</strong> parameters is set to <code>true</code> (default), the tool will perform annotation on the provided VCF file. This process involves running snpEff to enrich the VCF with annotation data by leveraging databases such as dbNFSP, ClinVar, and COSMIC. In this mode, the tool will generate and populate the ANN fields based on the analysis.</p>
<p>If the <strong>annotate</strong> parameters is iset to <code>false</code>, the tool assumes that the VCF file has already been annotated (i.e., it already contains the ANN fields). In this case, it will skip the annotation step and directly parse the existing ANN fields. If you choose this route, please make sure to modify the file <code>lib/internal/complete/config.yaml</code> with your own values.</p>
</li>
<li>
<p><strong>bff</strong></p>
<p>Location for the Beacon Friendly Format JSON files.</p>
</li>
<li>
<p><strong>center</strong></p>
<p>Experimental feature. Not used for now.</p>
</li>
<li>
<p><strong>datasetid</strong></p>
<p>An unique identifier for the dataset present in the input VCF. Default value is 'id_1'</p>
</li>
<li>
<p><strong>genome</strong></p>
<p>Your reference genome.</p>
<p>Accepted values: hg19, hg38 and hs37.</p>
<p>If you used GATKs GRCh37/b37 set it to hg19.</p>
<p>Not supported: NCBI36/hg18, NCBI35/hg17, NCBI34/hg16, hg15 and older.</p>
</li>
<li>
<p><strong>organism</strong></p>
<p>Experimental feature. Not used for now.</p>
</li>
<li>
<p><strong>bff2html</strong></p>
<p>Set bff2html to <code>true</code> to create HTML for the BFF Genomic Variations Browser.</p>
</li>
<li>
<p><strong>projectdir</strong></p>
<p>The prefix for dir name (e.g., 'cancer_sample_001'). Note that it can also contain a path (e.g., /workdir/cancer_sample_001).
 The script will automatically add an unique identifier to each job.</p>
</li>
<li>
<p><strong>technology</strong></p>
<p>Experimental feature. Not used for now.</p>
</li>
</ul>
<p><strong>Examples:</strong></p>
<div class="language-text highlight"><pre><span></span><code> $ bin/bff-tools vcf -i input.vcf.gz

 $ bin/bff-tools vcf -i input.vcf.gz -p param.yaml -projectdir-override beacon_exome_id_123456789

 $ bin/bff-tools load -p param_file  # MongoDB load only

 $ bin/bff-tools full -t 1 --i input.vcf.gz -p param_file  &gt; log 2&gt;&amp;1

 $ bin/bff-tools full -t 1 --i input.vcf.gz -p param_file -c config_file &gt; log 2&gt;&amp;1

 $ bin/bff-tools validate -i my_data.xlsx -o outdir

 $ nohup $path_to_beacon/bin/bff-tools full -i input.vcf.gz -verbose

 $ parallel &quot;bin/bff-tools vcf -t 1 -i chr{}.vcf.gz  &gt; chr{}.log 2&gt;&amp;1&quot; ::: {1..22} X Y
</code></pre></div>
<p><em>NB</em>: If you don't want colors in the output use the flag <code>--no-color</code>. If you did not use the flag and want to get rid off the colors in your printed log file use this command to parse ANSI colors:</p>
<div class="language-text highlight"><pre><span></span><code> perl -pe &#39;s/\x1b\[[0-9;]*[mG]//g&#39;
</code></pre></div>
<p><strong>Note:</strong> The script creates <code>log</code> files for all the processes. For instance, when running in <code>vcf</code> mode you can check via <code>tail -f</code> command:</p>
<div class="language-text highlight"><pre><span></span><code> $ tail -f &lt;your_job_id/vcf/run_vcf2bff.log
</code></pre></div>
<p>## WHAT IS THE BEACON FRIENDLY FORMAT (BFF)</p>
<p>The Beacon Friendly Format is a data exchange format consisting up to  7 JSON files (JSON arrays) that match the 7 schemas from <a href="https://docs.genomebeacons.org/schemas-md/analyses_defaultSchema/">Beacon v2 Models</a>.</p>
<p>Six files correspond to Metadata (<code>analyses.json,biosamples.json,cohorts.json,datasets.json,individuals.json,runs.json</code>) and one corresponds to variations (<code>genomicVariations.json</code>).</p>
<p>Normally, <code>bff-tools</code> script is used to create <code>genomicVariations</code> JSON file. The other 6 files are created with <a href="https://github.com/CNAG-Biomedical-Informatics/beacon2-cbi-tools/tree/main/utils/bff_validator">this utility</a> (part of the distribution). See instructions <a href="https://github.com/CNAG-Biomedical-Informatics/beacon2-cbi-tools/tree/main/utils/bff_validator/README.md">here</a>.</p>
<p>Once we have all seven files, then we can proceed to load the data into MongoDB.</p>
<p># COMMON ERRORS: SYMPTOMS AND TREATMENT</p>
<div class="language-text highlight"><pre><span></span><code> * Perl: 
         * Execution errors:
           - Error with YAML::XS
             Solution: Make sure the YAML (config.yaml or parameters file) is well formatted (e.g., space after param:&#39; &#39;).

 * Bash: 
         (Possible errors that can happen when the embeded Bash scripts are executed)
         * bcftools errors: bcftools is nit-picky about VCF fields and nomenclature of contigs/chromosomes in reference genome
                =&gt; Failed to execute: beacon_161855926405757/run_vcf2bff.sh
                   Please check this file beacon_161855926405757/run_vcf2bff.log
           - Error: 
                  # Running bcftools
                  [E::faidx_adjust_position] The sequence &quot;22&quot; was not found
             Solution: Make sure you have set the correct genome (e.g., hg19, hg38 or hs37) in parameters_file.
                       In this case bcftools was expecting to find 22 in the &lt;*.fa.gz&gt; file from reference genome, but found &#39;chr22&#39; instead.
                 Tips:
                      - hg{19,38} use &#39;chr&#39; in chromosome naming (e.g., chr1)
                      - hs37 does not use &#39;chr&#39; in chromosome naming (e.g., 1)

            - Error
                 # Running bcftools
                 INFO field IDREP only contains 1 field, expecting 2
              Solution: Please Fix VCF info field manually (or get rid of problematic fields with bcftools)
                        e.g., bcftools annotate -x INFO/IDREP input.vcf.gz | gzip &gt; output.vcf.gz
                              bcftools annotate -x INFO/MLEAC,INFO/MLEAF,FMT/AD,FMT/PL input.vcf.gz  | gzip &gt; output.vcf.gz

   NB: The bash scripts can be executed &quot;manually&quot; in the beacon_XXX dir. You must provide the 
       input vcf as an argument. This is a good option for debugging.
</code></pre></div>
<p>## KNOWN ISSUES</p>
<div class="language-text highlight"><pre><span></span><code> * Some Linux distributions do not include perldoc and thus Perl&#39;s library Pod::Usage will complain.
   Please, install it (sudo apt install perl-doc) if needed.
</code></pre></div>
<p># CITATION</p>
<p>The author requests that any published work that utilizes <strong>B2RI</strong> includes a cite to the the following reference:</p>
<p>Rueda, M, Ariosa R. "Beacon v2 Reference Implementation: a toolkit to enable federated sharing of genomic and phenotypic data". <em>Bioinformatics</em>, btac568, <a href="https://doi.org/10.1093/bioinformatics/btac568">doi.org/10.1093/bioinformatics/btac568</a></p>
<p># AUTHOR</p>
<p>Written by Manuel Rueda, PhD. Info about CNAG can be found at <a href="https://www.cnag.eu">https://www.cnag.eu</a></p>
<p>Credits: </p>
<div class="language-text highlight"><pre><span></span><code> * Sabela De La Torre (SDLT) created a Bash script for Beacon v1 to parse vcf files L&lt;https://github.com/ga4gh-beacon/beacon-elixir&gt;.
 * Toshiaki Katayamai re-implemented the Beacon v1 script in Ruby.
 * Later Dietmar Fernandez-Orth (DFO) modified the Ruby for Beacon v2 L&lt;https://github.com/ktym/vcftobeacon and added post-processing with R, from which I borrowed ideas to implement vcf2bff.pl.
 * DFO for usability suggestions and for creating bcftools/snpEff commands.
 * Roberto Ariosa for help with MongoDB implementation.
 * Mauricio Moldes helped with the containerization.
</code></pre></div>
<p># COPYRIGHT and LICENSE</p>
<p>This PERL file is copyrighted. See the LICENSE file included in this distribution.</p>
</div>
<div class="tabbed-block">
<p>For executing <code>bff-validator</code> you will need:</p>
<ul>
<li>
<p>Input file:</p>
<p>You have two options:</p>
<p><strong>A)</strong> A XLSX file consisting of multiple sheets. A <a href="https://metacpan.org/pod/Beacon-v2-Models_template.xlsx">template</a> version of this file is provided with this installation.</p>
<p>Currently, the file consists of 7 sheets that match the Beacon v2 Models.</p>
<p>Please use the flag <code>--gv</code> should you want to validate the data in the sheet &lt;genomicVariations&gt;.</p>
<p><em>NB:</em> If you have multiple CSV files instead of a XLSX file you can use the included utility <a href="https://github.com/EGA-archive/beacon2-cbi-tools/blob/main/utils/models2xlsx/csv2xlsx">csv2xlsx</a> that will join all CSVs into 1 XLSX.</p>
<div class="language-text highlight"><pre><span></span><code> $ ./csv2xlsx *csv -o out.xlsx
</code></pre></div>
<p><strong>B)</strong> A set of JSON (array) files that follow the Beacon Friendly Format. The files MUST be uncompressed and named &lt;analyses.json&gt;, &lt;biosamples.json&gt;, etc.</p>
</li>
<li>
<p>Beacon v2 Models (with JSON pointers dereferenced)</p>
<p>You should have them at <code>deref_schemas</code> directory.</p>
</li>
</ul>
<p><strong>Examples:</strong></p>
<div class="language-text highlight"><pre><span></span><code>  $ ./bff-validator -i file.xlsx

  $ $path/bff-validator -i file.xlsx -o my_bff_outdir

  $ $path/bff-validator -i my_bff_in_dir/*json -s deref_schemas -o my_bff_out_dir

  $ $path/bff-validator -i file.xlsx --gv --schema-dir deref_schemas --out-dir my_bff_out_dir
</code></pre></div>
<p>## TIPS ON FILLING OUT THE EXCEL TEMPLATE</p>
<div class="language-text highlight"><pre><span></span><code> * Please, before filling in any field, check out the provided template for ../../CINECA_synthetic_cohort_EUROPE_UK1/*xlsx
 * The script accepts Unicode characters (encoded with utf-8)
 * Header fields: 
    - Dots (&#39;.&#39;) represent objects: 
        Examples (values):
          1 - foo
          2 - NCIT:C20197
          3 - true # booleans
          4 - [&quot;foo&quot;,&quot;bar&quot;,&quot;baz&quot;] # arrays are also allowed
    - Underscores (&#39;_&#39;) represent arrays: 
        * Up to 1D (e.g., individuals-&gt;measures_assyCode.id) the values are comma separated
           Examples (values):
            1 - measures_assayCode.id
                LOINC:35925-4,LOINC:3141-9,LOINC:8308-9
               measures_assayCode.label
                BMI,Weight,Height-standing

        * Others - Values for array fields start with &#39;[&#39; and end with &#39;]&#39;
           Examples (values): 
            1 - [&quot;foo&quot;:{&quot;bar&quot;: &quot;baz&quot;}}]
            2 - [&quot;foo&quot;,&quot;bar&quot;,&quot;baz&quot;]
</code></pre></div>
<p>## COMMON ERRORS AND SOLUTIONS</p>
<div class="language-text highlight"><pre><span></span><code> * Error message: , or } expected while parsing object/hash, at character offset 574 (before &quot;]&quot;)
   Solution: Make sure you have the right amount of opening or closing keys/brackets.
</code></pre></div>
<p><em>NB:</em> You can use the flag <code>--ignore-validation</code> and check the temporary files at <code>-o</code> directory.</p>
</div>
</div>
</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; Copyright 2023-2025, Manuel Rueda, CNAG
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/mrueda" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.select", "content.code.copy", "content.tabs.link", "search.highlight", "search.share", "navigation.tracking", "navigation.sections", "toc.integrate"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../js/theme_overrides.js"></script>
      
    
  </body>
</html>